<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.26">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="Camilo García-Botero">
<meta name="dcterms.date" content="2021-02-25">
<title>Miscellaneous - K-mer counts on genomes</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<link href="../../site_libs/quarto-contrib/academicons-1.9.2/all.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/academicons-1.9.2/size.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
<meta name="citation_title" content="K-mer counts on genomes">
<meta name="citation_author" content="Camilo García-Botero">
<meta name="citation_publication_date" content="2021-02-25">
<meta name="citation_cover_date" content="2021-02-25">
<meta name="citation_year" content="2021">
<meta name="citation_online_date" content="2021-02-25">
<meta name="citation_language" content="en">
</head>
<body class="floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg navbar-dark "><div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Miscellaneous</span>
    </a>
  </div>
          <div class="quarto-toggle-container ms-auto">
              <a href="" class="quarto-color-scheme-toggle nav-link" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
          </div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav></header><!-- content --><header id="title-block-header" class="quarto-title-block default toc-left page-columns page-full"><div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">K-mer counts on genomes</h1>
                                <div class="quarto-categories">
                <div class="quarto-category">bioinformatics</div>
                <div class="quarto-category">code</div>
                <div class="quarto-category">analysis</div>
              </div>
                  </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p><a href="https://camilogarciabotero.github.io/">Camilo García-Botero</a> <a href="https://orcid.org/0000-0002-0426-7007" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">February 25, 2021</p>
      </div>
    </div>
    
      <div>
      <div class="quarto-title-meta-heading">Modified</div>
      <div class="quarto-title-meta-contents">
        <p class="date-modified">November 23, 2022</p>
      </div>
    </div>
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto"><nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">On this page</h2>
   
  <ul>
<li><a href="#libraries" id="toc-libraries" class="nav-link active" data-scroll-target="#libraries">Libraries</a></li>
  <li><a href="#downloading-genomes" id="toc-downloading-genomes" class="nav-link" data-scroll-target="#downloading-genomes">Downloading genomes</a></li>
  <li><a href="#filenames-cleaning" id="toc-filenames-cleaning" class="nav-link" data-scroll-target="#filenames-cleaning">Filenames cleaning</a></li>
  <li>
<a href="#k-mers-scanning-on-genomes" id="toc-k-mers-scanning-on-genomes" class="nav-link" data-scroll-target="#k-mers-scanning-on-genomes">K-mers scanning on genomes</a>
  <ul class="collapse">
<li><a href="#script-help-parallel-execution" id="toc-script-help-parallel-execution" class="nav-link" data-scroll-target="#script-help-parallel-execution">Script help &amp; parallel execution</a></li>
  </ul>
</li>
  <li><a href="#importing-cleaning" id="toc-importing-cleaning" class="nav-link" data-scroll-target="#importing-cleaning">Importing &amp; cleaning</a></li>
  <li><a href="#k-mers-counts-frequency-plot" id="toc-k-mers-counts-frequency-plot" class="nav-link" data-scroll-target="#k-mers-counts-frequency-plot">K-mers counts frequency plot</a></li>
  </ul><div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/camilogarciabotero/blog/edit/main/posts/2021-01-25-Kmer-analysis/kmer_counts.qmd" class="toc-action">Edit this page</a></p><p><a href="https://github.com/camilogarciabotero/blog/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
</nav><!-- margin-sidebar --><div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content"><section id="libraries" class="level1"><h1>Libraries</h1>
<div class="cell" data-hash="kmer_counts_cache/html/libraries_ca79fec118cc58ae3a56c8bccd9a3486">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://fs.r-lib.org">fs</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://rstudio.github.io/reticulate/">reticulate</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://csdaw.github.io/ggprism/">ggprism</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/slowkow/ggrepel">ggrepel</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com">data.table</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://yihui.org/knitr/">knitr</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="downloading-genomes" class="level1"><h1>Downloading genomes</h1>
<p>The first steps of this challenge requires genome data. A very practical way to download data programmatically is to use <code>ncbi-genome-downlad</code> <a href="">program</a>. Here a you can see how to use it:</p>
<p>Then using we will download complete genomes from the <em>Pseudomonas syringae</em> and drop them into the <strong>Genome</strong> folder. In the following line we select ~37 genomes from the organism.</p>
<div class="cell" data-hash="kmer_counts_cache/html/unnamed-chunk-1_890d069cb98ac4e6532e8ebd84c29f0c">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode zsh code-with-copy"><code class="sourceCode zsh"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">ngd</span> <span class="at">-n</span> <span class="at">-s</span> refseq <span class="at">-F</span> fasta <span class="at">--genera</span> <span class="st">"Pseudomonas syringae"</span> <span class="at">-l</span> complete bacteria</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Considering the following 47 assemblies for download:
GCF_004006335.1 Pseudomonas syringae    inb918
GCF_014524645.1 Pseudomonas syringae    CAS02
GCF_016694755.2 Pseudomonas syringae    BIM B-268
GCF_018388485.1 Pseudomonas syringae    KF529
GCF_018388505.1 Pseudomonas syringae    Susan762
GCF_018388525.1 Pseudomonas syringae    U643
GCF_018394375.1 Pseudomonas syringae    Susan2139
GCF_023278085.1 Pseudomonas syringae    PA-2-9E
GCF_900235815.1 Pseudomonas syringae    CFBP3840
GCF_900289125.1 Pseudomonas syringae    CFBP 2116
GCF_002905815.2 Pseudomonas syringae pv. syringae   Pss9097
GCF_023277945.1 Pseudomonas syringae pv. syringae   Pss9644
GCF_900235825.1 Pseudomonas syringae pv. syringae   CFBP4215
GCF_900235865.1 Pseudomonas syringae pv. syringae   CFBP2118
GCF_002966555.1 Pseudomonas syringae pv. tomato B13-200
GCF_009800225.1 Pseudomonas syringae pv. tomato delta X
GCF_016599655.1 Pseudomonas syringae pv. maculicola MAFF 302723
GCF_001913215.1 Pseudomonas syringae pv. actinidiae ICMP 20586
GCF_001913235.1 Pseudomonas syringae pv. actinidiae NZ-47
GCF_002024285.1 Pseudomonas syringae pv. actinidiae CRAFRU 12.29
GCF_002024305.1 Pseudomonas syringae pv. actinidiae CRAFRU 14.08
GCF_002763655.1 Pseudomonas syringae pv. actinidiae MAFF212063
GCF_003665415.1 Pseudomonas syringae pv. actinidiae P220
GCF_023167685.1 Pseudomonas syringae pv. actinidiae MAFF613020
GCF_022557255.1 Pseudomonas syringae pv. tagetis    ICMP 4091
GCF_003047185.1 Pseudomonas syringae pv. atrofaciens    LMG5095
GCF_001482725.1 Pseudomonas syringae pv. lapsa  ATCC 10859
GCF_000012245.1 Pseudomonas syringae pv. syringae B728a B728a
GCF_000007805.1 Pseudomonas syringae pv. tomato str. DC3000 DC3000
GCF_022557235.1 Pseudomonas syringae pv. helianthi  LMG 5067
GCF_900235905.1 Pseudomonas syringae group genomosp. 3  CFBP6411
GCF_900235885.1 Pseudomonas syringae pv. cerasicola CFBP6109
GCF_000145825.2 Pseudomonas syringae Cit 7  Cit 7
GCF_000145845.2 Pseudomonas syringae pv. maculicola str. ES4326 ES4326
GCF_900235835.1 Pseudomonas syringae pv. avii   CFBP3846
GCF_000648735.3 Pseudomonas syringae pv. actinidiae ICMP 18884  ICMP 18884
GCF_000344335.2 Pseudomonas syringae pv. actinidiae ICMP 9853   ICMP 9853
GCF_000344355.2 Pseudomonas syringae pv. actinidiae ICMP 18708  ICMP 18708
GCF_000344475.3 Pseudomonas syringae pv. actinidiae str. Shaanxi_M228   Shaanxi_M228
GCF_000988485.1 Pseudomonas syringae pv. syringae B301D B301D
GCF_000988395.1 Pseudomonas syringae pv. syringae HS191 HS191
GCF_001281365.1 Pseudomonas syringae UMAF0158   UMAF0158
GCF_000452705.1 Pseudomonas syringae CC1557 CC1557
GCF_000452605.2 Pseudomonas syringae CC440  CC440
GCF_000452565.2 Pseudomonas syringae UB303  UB303
GCF_000452525.3 Pseudomonas syringae USA011 USA011
GCF_000452445.2 Pseudomonas syringae pv. pisi str. PP1  PP1</code></pre>
</div>
</div>
<div class="cell" data-hash="kmer_counts_cache/html/downloading_ngd_a4640b80e266f1f85cf2bcf61a0e3e5a">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode zsh code-with-copy"><code class="sourceCode zsh"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">ngd</span> <span class="at">-s</span> refseq<span class="dt">\ </span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>    <span class="ex">-F</span> fasta<span class="dt">\</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">--genera</span> <span class="st">"Pseudomonas syringae"</span><span class="dt">\</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">-l</span> complete<span class="dt">\</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">--flat-output</span> bacteria<span class="dt">\</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">-o</span> Data/Genomes</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="fu">gzip</span> <span class="at">-d</span> Data/Genomes/<span class="pp">*</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="filenames-cleaning" class="level1"><h1>Filenames cleaning</h1>
<div class="cell" data-hash="kmer_counts_cache/html/unnamed-chunk-3_0e7e8da121a55593bb1cdf9c61b89270">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode zsh code-with-copy"><code class="sourceCode zsh"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">python</span> ~/Programs/Bioinf_tools/bit-dedupe-fasta-headers.py <span class="at">-h</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>usage: bit-dedupe-fasta-headers.py [-h] -i INPUT_FASTA [-o OUTPUT_FASTA_NAME]

This script will append a number to headers if that exact ID has already
appeared in the fasta file. For version info, run `bit-version`.

optional arguments:
  -h, --help            show this help message and exit
  -o OUTPUT_FASTA_NAME, --output_fasta_name OUTPUT_FASTA_NAME
                        Output fasta file (default: "Renamed.fasta").

required arguments:
  -i INPUT_FASTA, --input_fasta INPUT_FASTA
                        Starting fasta file</code></pre>
</div>
</div>
<div class="cell" data-hash="kmer_counts_cache/html/unnamed-chunk-4_8f5711b8a8b0684eaecd4bc08c025859">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode zsh code-with-copy"><code class="sourceCode zsh"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> F <span class="kw">in</span> Genomes/<span class="pp">*</span>.fna <span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>    <span class="va">N</span><span class="op">=</span><span class="va">$(</span><span class="fu">basename</span> <span class="va">$F</span> .fna<span class="va">)</span> <span class="kw">;</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>    <span class="ex">python</span> /Users/camilogarcia/Programs/Bioinf_tools/bit-dedupe-fasta-headers.py <span class="at">-i</span> <span class="va">$F</span> <span class="at">-o</span> <span class="va">$N_renamed</span>.mfa <span class="kw">;</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="kmer_counts_cache/html/unnamed-chunk-5_68cde609b2b032fd4ef9ffe5edfe2b59">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode zsh code-with-copy"><code class="sourceCode zsh"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> F <span class="kw">in</span> Genomes/<span class="pp">*</span>.fna<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mv</span> <span class="at">--</span> <span class="st">"</span><span class="va">$F</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"</span><span class="va">$(</span><span class="fu">awk</span> <span class="st">'NR==1{printf("%s_%s_%s\n",$2,$3,substr($1,2));exit}'</span> <span class="st">"</span><span class="va">$F</span><span class="st">"</span><span class="va">)</span><span class="st">"</span>.fna</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="k-mers-scanning-on-genomes" class="level1"><h1>K-mers scanning on genomes</h1>
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><a href="https://github.com/camilogarciabotero/Biolibrary"><img src="imgs/carbon_kmer_count.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
<section id="script-help-parallel-execution" class="level2"><h2 class="anchored" data-anchor-id="script-help-parallel-execution">Script help &amp; parallel execution</h2>
<div class="cell" data-hash="kmer_counts_cache/html/unnamed-chunk-6_c085ea7392aee3ccb3eea1c7128bd651">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode zsh code-with-copy"><code class="sourceCode zsh"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="ex">python</span> ~/Projects/Biolibrary/kmer_count.py <span class="at">-h</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>usage: kmer_count.py [-h] [-k K] input output

Count the frecuency of a k-mer set given its size (k) along a genome or a set
of genomes

positional arguments:
  input       Path to input fasta file
  output      Path to put file/folder output

optional arguments:
  -h, --help  show this help message and exit
  -k K        size of k-mer (default: 1)</code></pre>
</div>
</div>
<div class="cell" data-hash="kmer_counts_cache/html/unnamed-chunk-7_cbfecebc44cff02db8407cbae1a755aa">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode zsh code-with-copy"><code class="sourceCode zsh"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="ex">l</span> ~/Projects/Random/Genomes/Pseudomonas_syringae_NZ<span class="pp">*</span> <span class="kw">|</span> <span class="ex">parallel</span> <span class="st">"python kmer_count.py -k 2 {} {/.}.csv"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="kmer_counts_cache/html/unnamed-chunk-8_9c441fc434b92445849a95a35c9b5386">
<div class="cell-output cell-output-stdout">
<pre><code>Pseudomonas_syringae_NZ_CP005969.1_9mers.csv
Pseudomonas_syringae_NZ_CP005970.1_9mers.csv
Pseudomonas_syringae_NZ_CP006256.1_9mers.csv
Pseudomonas_syringae_NZ_CP007014.1_9mers.csv
Pseudomonas_syringae_NZ_CP011972.2_9mers.csv
Pseudomonas_syringae_NZ_CP012179.1_9mers.csv
Pseudomonas_syringae_NZ_CP013183.1_9mers.csv
Pseudomonas_syringae_NZ_CP017007.1_9mers.csv
Pseudomonas_syringae_NZ_CP017009.1_9mers.csv
Pseudomonas_syringae_NZ_CP018202.1_9mers.csv
Pseudomonas_syringae_NZ_CP019730.1_9mers.csv
Pseudomonas_syringae_NZ_CP019732.1_9mers.csv
Pseudomonas_syringae_NZ_CP019871.1_9mers.csv
Pseudomonas_syringae_NZ_CP024646.1_9mers.csv
Pseudomonas_syringae_NZ_CP024712.1_9mers.csv
Pseudomonas_syringae_NZ_CP026568.1_9mers.csv
Pseudomonas_syringae_NZ_CP028490.1_9mers.csv
Pseudomonas_syringae_NZ_CP032459.1_9mers.csv
Pseudomonas_syringae_NZ_CP032631.1_9mers.csv
Pseudomonas_syringae_NZ_CP032871.1_9mers.csv
Pseudomonas_syringae_NZ_CP034078.1_9mers.csv
Pseudomonas_syringae_NZ_CP045799.1_9mers.csv
Pseudomonas_syringae_NZ_CP047073.1_9mers.csv
Pseudomonas_syringae_NZ_CP047260.1_9mers.csv
Pseudomonas_syringae_NZ_CP047267.1_9mers.csv
Pseudomonas_syringae_NZ_CP067024.1_9mers.csv
Pseudomonas_syringae_NZ_CP068034.1_9mers.csv
Pseudomonas_syringae_NZ_LT962480.1_9mers.csv
Pseudomonas_syringae_NZ_LT962481.1_9mers.csv
Pseudomonas_syringae_NZ_LT963391.1_9mers.csv
Pseudomonas_syringae_NZ_LT963402.1_9mers.csv
Pseudomonas_syringae_NZ_LT963408.1_9mers.csv
Pseudomonas_syringae_NZ_LT963409.1_9mers.csv
Pseudomonas_syringae_NZ_LT985192.1_9mers.csv</code></pre>
</div>
</div>
</section></section><section id="importing-cleaning" class="level1"><h1>Importing &amp; cleaning</h1>
<div class="cell" data-hash="kmer_counts_cache/html/unnamed-chunk-9_7c84704a32e2d81df5badb58c86ecd7e">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># posts/2021-01-25-Kmer-analysis/</span></span>
<span></span>
<span><span class="va">files</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://fs.r-lib.org/reference/dir_ls.html">dir_ls</a></span><span class="op">(</span><span class="st">"data/Kmers"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dataset</span>  <span class="op">&lt;-</span> <span class="va">files</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map_df</a></span><span class="op">(</span><span class="va">read_csv</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dataset</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 20,709,376 × 3
    ...1 kmer      kmer_value
   &lt;dbl&gt; &lt;chr&gt;          &lt;dbl&gt;
 1     0 AAAAAAAAA          5
 2     1 AAAAAAAAC         21
 3     2 AAAAAAAAT         17
 4     3 AAAAAAAAG         17
 5     4 AAAAAAACA         23
 6     5 AAAAAAACC         44
 7     6 AAAAAAACT         15
 8     7 AAAAAAACG         29
 9     8 AAAAAAATA         26
10     9 AAAAAAATC         40
# … with 20,709,366 more rows</code></pre>
</div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">total</span> <span class="op">&lt;-</span> <span class="va">dataset</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>kmer <span class="op">=</span> <span class="fu"><a href="https://forcats.tidyverse.org/reference/as_factor.html">as_factor</a></span><span class="op">(</span><span class="va">kmer</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">kmer</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span>total <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">kmer_value</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">total</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped">
<thead><tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">kmer</th>
<th style="text-align: left;">total</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"></td>
<td style="text-align: left;">AAAAAAAAA: 1</td>
<td style="text-align: left;">Min. : 3</td>
</tr>
<tr class="even">
<td style="text-align: left;"></td>
<td style="text-align: left;">AAAAAAAAC: 1</td>
<td style="text-align: left;">1st Qu.: 235</td>
</tr>
<tr class="odd">
<td style="text-align: left;"></td>
<td style="text-align: left;">AAAAAAAAT: 1</td>
<td style="text-align: left;">Median : 509</td>
</tr>
<tr class="even">
<td style="text-align: left;"></td>
<td style="text-align: left;">AAAAAAAAG: 1</td>
<td style="text-align: left;">Mean : 819</td>
</tr>
<tr class="odd">
<td style="text-align: left;"></td>
<td style="text-align: left;">AAAAAAACA: 1</td>
<td style="text-align: left;">3rd Qu.: 1047</td>
</tr>
<tr class="even">
<td style="text-align: left;"></td>
<td style="text-align: left;">AAAAAAACC: 1</td>
<td style="text-align: left;">Max. :15920</td>
</tr>
<tr class="odd">
<td style="text-align: left;"></td>
<td style="text-align: left;">(Other) :262138</td>
<td style="text-align: left;">NA</td>
</tr>
</tbody>
</table>
</div>
</div>
</section><section id="k-mers-counts-frequency-plot" class="level1"><h1>K-mers counts frequency plot</h1>
<div class="cell" data-hash="kmer_counts_cache/html/kmer-plot_d850341bf8017f4ef3ed20bf0ee4b730">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">kmers_plot</span> <span class="op">&lt;-</span> <span class="va">total</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">total</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"count"</span>, color <span class="op">=</span> <span class="st">"#8A8A8A"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html">geom_label_repel</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">kmer</span>, <span class="st">"appears"</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">total</span><span class="op">)</span>, <span class="st">"\n times in all genomes"</span><span class="op">)</span>, y <span class="op">=</span> <span class="fl">120</span><span class="op">)</span>, data <span class="op">=</span> <span class="va">.</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">total</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">total</span><span class="op">)</span><span class="op">)</span>, min.segment.length <span class="op">=</span> <span class="fl">0</span>, segment.colour <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">total</span><span class="op">$</span><span class="va">total</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span>, color <span class="op">=</span> <span class="st">"#C60000"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotate.html">annotate</a></span><span class="op">(</span><span class="st">"label"</span>, label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html">expression</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">mu</span>, <span class="st">"="</span>, <span class="fl">818</span><span class="op">)</span><span class="op">)</span>, x <span class="op">=</span> <span class="fl">1100</span>, y <span class="op">=</span> <span class="fl">350</span>, size <span class="op">=</span> <span class="fl">4</span>, colour <span class="op">=</span> <span class="st">"#C60000"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span><span class="op">(</span>guide <span class="op">=</span> <span class="st">"prism_offset_minor"</span>, trans <span class="op">=</span> <span class="st">"log10"</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">18000</span><span class="op">)</span>, n.breaks <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span></span>
<span>    axis.line.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_line</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="st">"K-mer value [n]"</span>,</span>
<span>    y <span class="op">=</span> <span class="st">"Absolute frequency [n]"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html">coord_cartesian</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="va">kmers_plot_intervals</span> <span class="op">&lt;-</span> <span class="va">total</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>intervals <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cut.html">cut</a></span><span class="op">(</span><span class="va">total</span>, breaks <span class="op">=</span> <span class="fl">19</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">intervals</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"count"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span><span class="op">(</span>guide <span class="op">=</span> <span class="st">"prism_offset_minor"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span></span>
<span>    axis.line.x  <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_line</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="st">"Absolute frequency [n]"</span>,</span>
<span>    y <span class="op">=</span> <span class="st">""</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html">coord_cartesian</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="co">#   ggsave("kmers_freq_intervals_01.png", width = 10, height = 10, dpi = 400)</span></span>
<span> </span>
<span></span>
<span><span class="va">kmers_plot</span> <span class="op">/</span> <span class="va">kmers_plot_intervals</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="kmer_counts_files/figure-html/kmer-plot-1.png" class="img-fluid" width="3000"></p>
</div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#   ggsave("kmers_freq_plots_01.png", width = 10, height = 12, dpi = 400)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


</section><div id="quarto-appendix" class="default"><section class="quarto-appendix-contents"><h2 class="anchored quarto-appendix-heading">Citation</h2><div><div class="quarto-appendix-secondary-label">BibTeX citation:</div><pre class="sourceCode code-with-copy quarto-appendix-bibtex"><code class="sourceCode bibtex">@online{garcía-botero2021,
  author = {Camilo García-Botero},
  title = {K-Mer Counts on Genomes},
  date = {2021-02-25},
  langid = {en}
}
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre><div class="quarto-appendix-secondary-label">For attribution, please cite this work as:</div><div id="ref-garcía-botero2021" class="csl-entry quarto-appendix-citeas" role="doc-biblioentry">
Camilo García-Botero. 2021. <span>“K-Mer Counts on Genomes.”</span>
February 25, 2021.
</div></div></section></div></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><script src="https://giscus.app/client.js" data-repo="camilogarciabotero/blog" data-repo-id="" data-category="Blog" data-category-id="" data-mapping="title" data-reactions-enabled="1" data-emit-metadata="0" data-input-position="top" data-theme="transparent_dark" data-lang="en" crossorigin="anonymous" async="">
</script>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
      <div class="nav-footer-center">
        <ul class="footer-items list-unstyled">
<li class="nav-item compact">
    <a class="nav-link" href="https://github.com/camilogarciabotero">
      <i class="bi bi-github" role="img">
</i> 
    </a>
  </li>  
    <li class="nav-item">
    <a class="nav-link" href="https://scholar.google.com/citations?user=OJ6452UAAAAJ&amp;hl=en"><i class="ai  ai-google-scholar"></i></a>
  </li>  
    <li class="nav-item">
    <a class="nav-link" href="https://orcid.org/0000-0002-0426-7007"><i class="fa-brands fa-orcid" aria-label="orcid"></i></a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/gaspardelanoche">
      <i class="bi bi-twitter" role="img">
</i> 
    </a>
  </li>  
    <li class="nav-item">
    <a class="nav-link" href="https://camilogarciabotero.github.io/"><i class="fa-solid fa-house" aria-label="house"></i></a>
  </li>  
    <li class="nav-item">
    <a class="nav-link" href="https://www.paypal.com/paypalme/camilogarciabotero"><i class="fa-solid fa-mug-saucer" aria-label="mug-saucer"></i></a>
  </li>  
</ul>
</div>
  </div>
</footer>


</body></html>